webserver/views/datasets.py

```diff
@@ -286,36 +298,65 @@ def create_service():
 def import_csv():
     form = forms.DatasetCSVImportForm()
     if form.validate_on_submit():
+        description, classes = _parse_dataset_csv(request.files[form.file.name])
         dataset_dict = {
             "name": form.name.data,
-            "description": form.description.data,
-            "classes": _parse_dataset_csv(request.files[form.file.name]),
+            "description": description if description else form.description.data,
+            "classes": classes,
             "public": True,
         }
         try:
             dataset_id = db.dataset.create_from_dict(dataset_dict, current_user.id)
         except dataset_validator.ValidationException as e:
             raise BadRequest(str(e))
         flash.info("Dataset has been imported successfully.")
-        return redirect(url_for(".view", id=dataset_id))
+        return redirect(url_for(".view", dataset_id=dataset_id))
 
     else:
         return render_template("datasets/import.html", form=form)
 
 
 def _parse_dataset_csv(file):
-    classes_dict = defaultdict(list)
+    """Parse a csv file containing a representation of a dataset.
+    The csv file should have rows with 2 columns in one of the following forms:
+      <recording_id>,<classname>
+      description,<dataset_description>
+      description:<classname>,<class_description>
+
+    Arguments:
+        file: path to the csv file containing the dataset
+    Returns: a tuple of (dataset description, [classes]), where classes is a list of dictionaries
+             {"name": class name, "description": class description, "recordings": []}
+             a class is only returned if there are recordings for it. A class
+        """
+    classes_dict = defaultdict(lambda: {"description": None, "recordings": []})
+    dataset_description = None
     for class_row in csv.reader(file):
         if len(class_row) != 2:
             raise BadRequest("Bad dataset! Each row must contain one <MBID, class name> pair.")
-        classes_dict[class_row[1]].append(class_row[0])
+
+        if class_row[0] == "description":
+            # row is the dataset description
+            dataset_description = class_row[1]
+        elif (class_row[0])[:12] == "description:":
+            # row is a class description
+            class_name = class_row[0][12:]
+            classes_dict[class_name]["description"] = class_row[1]
+        else:
+            # row is a recording
+            classes_dict[class_row[1]]["recordings"].append(class_row[0])
+    
     classes = []
-    for name, recordings in six.iteritems(classes_dict):
-        classes.append({
-            "name": name,
-            "recordings": recordings,
-        })
-    return classes
+    
+    for name, class_data in six.iteritems(classes_dict):
+        if class_data["recordings"]:
```

Comment:
Thanks, I hadn't properly considered there being a minimum on the number of recordings. I have a couple of follow-up questions. With the minimum being two recordings, should we check for `class_data["recordings"]` to contain at least two recordings, rather than just checking that it isn't empty? Additionally, since we have this check, need we check below in `classes.append` for `"recordings" in class_data`?

Reply:
No, we don't need this check because during the upload process we call `db.dataset.create_from_dict`, which runs dataset validation first: https://github.com/metabrainz/acousticbrainz-server/blob/4c8e5d49d3b05d3d95831ba1c80254092b3a3889/db/dataset.py#L34
This will show an error on the upload page if the file doesn't have enough recordings

Reply:
Understood. Thanks, again!